### PR TITLE
Add pytorch MNIST Workflow tutorial with FedProx

### PR DESCRIPTION
**Pull request by ynonflumintel**
This branch introduces a tutorial for setting up a neural network training federation using pytorch, the MNIST dataset using the openfl workflow API.

### CODE REVIEW

**Comment by rahulga1** [comment]
@ynonflumintel  I think this is not the correct file in this commit, I dont see workflow api getting used in this file. 
can you please confirm?

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb [hunk 1]
```diff
@@ -0,0 +1,632 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    " #Get the current values of the plan. Each of these can be overridden\n",
+    "import json\n",
+    "print(json.dumps(fx.get_plan(), indent=4, sort_keys=True))"
```
Comment (teoparvanov) [inline openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb]: AFAIK, there's no notion of FL plan in Workflow API. Maybe consider dropping this cell?

Comment (ynonflumintel) [reply openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb]: I pushed with some uncommitted changes, it's fixed now

**Comment by ynonflumintel** [comment]
> @ynonflumintel I think this is not the correct file in this commit, I dont see workflow api getting used in this file. can you please confirm?

I pushed the wrong version of the file before, please take a look now

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb [hunk 2]
```diff
@@ -0,0 +1,343 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Federated FedProx PyTorch MNIST Workflow API Tutorial\n",
+    "This notebook mimics Federated_FedProx_PyTorch_MNIST_Tutorial using the workflow API.\n"
```
Comment (teoparvanov) [inline openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb]: I don't think we need to reference the deprecated (and soon to be deleted legacy notebook). Instead, I'd suggest providing a short overview of the new notebook (FedProx algo, Workflow API, LocalRuntime - scalable to a federated setting in the future).

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+    "        self.model = model\n",
+    "        self.optimizer = optimizer\n",
+    "        self.rounds = rounds\n",
+    "        self.loss = 0.\n",
+    "\n",
+    "    @aggregator\n",
+    "    def start(self):\n",
+    "        print(f'Performing initialization for model')\n",
+    "        self.collaborators = self.runtime.collaborators\n",
+    "        self.private = 10\n",
```
Comment (teoparvanov) [inline openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb]: The `private` field isn't really needed here. It's there in the 101 tutorial to illustrate how certain attributes can be explicitly filtered out when switching to the next step of the flow.

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+    "        self.current_round = 0\n",
+    "        self.next(self.aggregated_model_validation, foreach='collaborators', exclude=['private'])\n",
+    "\n",
+    "    def compute_accuracy(self, data_loader):\n",
```
Comment (teoparvanov) [inline openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb]: For improved readability of the federated flow spec, I suggest to keep utility methods towards the bottom.

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb
Comment (teoparvanov) [inline openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb:1]: Let's move this notebook together with the other experimental/workflow-based ones. Also, consider prefixing it as `105_`, to follow the "curriculum module" numbering pattern. Alternatively, we could consider prefixing it as `401_`, and renaming the non-IID FedProx notebook as `402_`. WDYT?

**Review by teoparvanov — APPROVED** [review]
Looks great, thanks @ynonflumintel ! I just have a couple of small suggestions:

**Review by teoparvanov — COMMENTED** [review]
PS: don't forget to sign off your commits for a successful CI run (DCO job):
```shell
git commit -s -m "..."
```

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb [hunk 5]
```diff
@@ -0,0 +1,343 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Federated FedProx PyTorch MNIST Workflow API Tutorial\n",
```
Comment (MasterSkepticista) [inline openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb]: ~Federated~ FedProx PyTorch MNIST Tutorial using Workflow API

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+    "import torch.utils.data\n",
+    "import torchvision\n",
+    "import torchvision.transforms as transforms\n",
+    "\n",
+    "from openfl.utilities.optimizers.torch.fedprox import FedProxAdam\n",
+    "\n",
+    "from openfl.experimental.interface import FLSpec, Aggregator, Collaborator\n",
+    "from openfl.experimental.runtime import LocalRuntime\n",
+    "from openfl.experimental.placement import aggregator, collaborator\n",
+    "\n",
```
Comment (MasterSkepticista) [inline openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb]: Suggest splitting the cell here. Helps partition the context. The functions in this cell should be close to where they are actually called.

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+    "        self.model.train()\n",
+    "        self.optimizer = get_optimizer(self.model)\n",
+    "        for batch_idx, (data, target) in enumerate(self.train_loader):\n",
+    "            self.optimizer.zero_grad()\n",
+    "            output = self.model(data)\n",
+    "            loss = F.cross_entropy(output, target)\n",
+    "            loss.backward()\n",
+    "            self.optimizer.step()\n",
+    "\n",
+    "            if (len(data) * batch_idx) / len(self.train_loader.dataset) >= log_threshold:\n",
```
Comment (MasterSkepticista) [inline openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb]: Instead of this complicated logic of calculating "quarter" of samples used, it is OK to print every 10 or 100 steps. We need to be cautious of all the cognitive overhead we put on the user.

**Review by MasterSkepticista — CHANGES_REQUESTED** [review]
General comment on the entire notebook: Include output of all the cells. New users should have visibility into the cells and evaluate for themselves if the notebook fulfils their goal of visit. Not showing output, can put-off newcomers because they end goal is not "visible".

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb (hunk 2 above)
Comment (ynonflumintel) [reply openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb]: I'm not using Fedprox as the aggregation algorithm here. Is it implemented somewhere and can be integrated here?

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb (hunk 3 above)
Comment (ynonflumintel) [reply openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb]: So if you remove the original and only keep this notebook, don't you want it here?

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb (hunk 4 above)
Comment (ynonflumintel) [reply openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb]: you need them defined before you use them so that you can use "run above" in the notebook and know it includes all of the dependencies

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb
Comment (ynonflumintel) [reply openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb:1]: I don't mind adding a prefix to the file name, which one do you want?
Where do you want this file to be moved to?
What does IID stand for?

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb (hunk 3 above)
Comment (teoparvanov) [reply openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb]: It's still available in he [101_MNIST](https://github.com/securefederatedai/openfl/blob/develop/openfl-tutorials/experimental/101_MNIST.ipynb) notebook, which is not going anywhere :)

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb (hunk 5 above)
Comment (ynonflumintel) [reply openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb]: done

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb (hunk 6 above)
Comment (ynonflumintel) [reply openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb]: done

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb
Comment (teoparvanov) [reply openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb:1]: I'd suggest `401_`, but currently this prefix is used for the "non-IID" notebook. You can rename the non-IID one to `402_`, and use the `401_` prefix for your notebook. 

Regarding IID, it means "Independent and Identically Distributed", i.e. all elements of the data set are sampled from the same statistical distribution. Non-IID means that dataset elements come from different distributions, which could be an obstacle for the convergence of the "vanilla" FedAvg. Enter FedProx, which offers better convergence for non-IID datasets (among other benefits related to dataset size and compute power heterogeneity among collaborators). You can read more about it [here](https://arxiv.org/pdf/1812.06127.pdf).

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb (hunk 7 above)
Comment (ynonflumintel) [reply openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb]: I ran experiments where there were a couple of thousand steps (small chunks), it's too spammy IMO

**Comment by ynonflumintel** [comment]
> General comment on the entire notebook: Include output of all the cells. New users should have visibility into the cells and evaluate for themselves if the notebook fulfils their goal of visit. Not showing output, can put-off newcomers because they end goal is not "visible".

Thought it would be too verbose and increase the file size significantly, but I added them as per your request

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb
Comment (ynonflumintel) [reply openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb:1]: please be specific, which file do you want me to name what - I'm getting a little lost

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb (hunk 3 above)
Comment (ynonflumintel) [reply openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb]: Oh, okay, done

File: openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb (hunk 2 above)
Comment (teoparvanov) [reply openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb]: Actually FedProx is not a "pure" aggregation algorithm, but a combination of the FedProx optimizer (which adds the so called "proximal" element to the loss function), and FedAvg for the aggregation part.

**Review by teoparvanov — APPROVED** [review]
